### ISSUE TITLE
make: Recursive make considered harmful

### ISSUE THREAD

**Issue by miri64** [RIOT-OS/RIOT#1242]
After reading through the article [Recursive make considered harmful](http://miller.emu.id.au/pmiller/books/rmch/) I realized that many of the problems we have with make are mentioned in it. Maybe we should consider the solutions suggested by it. ;-)

This might be a duplicate of several issues posted before. Sorry if this is the case.

**Comment by LudwigOrtmann** [comment]
@OlegHahm is your tab still open?

**Comment by OlegHahm** [comment]
Probably.

**Comment by pyropeter** [comment]
The article linked to in the issue description is now available here: http://aegis.sourceforge.net/auug97.pdf

**Comment by kaspar030** [comment]
I propose closing this. The build system is OK as is. Should we rewrite it, then definitely not using make again.

**Comment by OlegHahm** [comment]
I agree to the first sentence. I agree to the second if you define "OK" as "it works". But I disagree with the last sentence.

@miri64, fine with closing?

**Comment by pyropeter** [comment]
@nmeum and I are trying to add Rust support to RIOT at the moment. The ugly build system is a major issue.

**Comment by kaspar030** [comment]
> The ugly build system is a major issue.

Because it is using recursive make?

**Comment by OlegHahm** [comment]
I've never seen a non-ugly build system.

**Comment by miri64** [comment]
Mhhhh... I still think there could be some merit in straightening out the current build system... So no: would not close.

**Comment by pyropeter** [comment]
Building C code is comperativly easy because you can just compile each source file to a corresponding object file and then link everything as the last step of the build (this is what the current build system does, basically).

For rust support we need to completely build and link all dependencies of a module/package before we can build the module/package. This is impossible to express in the current build system without loosing the ability to have parallel builds.

**Comment by Kijewski** [comment]
There [are examples](https://www.google.de/search?q=rust+autotools) how to integrate rust with autotools, e.g. https://github.com/lu-zero/autotools-cargo-example. This could be a good starting point to integrate rust into RIOT.

That being said: It's clearly not out make system's fault that rust wants to reinvent the wheel …

**Comment by cladmi** [comment]
I started looking into this a few days ago. This comment is where I currently am, I need to write it down to help me settle things and also to get remarks.

From what I found now, to build applications, without taking the `gnrc_%` and `log_%` inclusion, the `pkg`, `unittest`, `BINDIRS`, `PSEUDOMODULES`, `SUBMODULES` into account, as I did not look into them, it may not be that hard to get rid of recursive makefiles.

The steps I see:
 * Add static definitions of MODULE_DIR_$(module) = "$(RIOTBASE)/relative_path_to_module"
    * This would allow replacing the dependency from `${DIRS:%=ALL--%}` to the real module file `$(BINDIR)/$(MODULE).a` even if they would be `.PHONY` at first
* Replace the `make -C ${dir}` by include `${dir}/Makefile and change `Makefile.base` to prepend `${dir}` to $(SRC) variables
* Then, the build targets would need to use $(CFLAGS_$(MODULE)) target. As `sys/crypto/Makefile` and `sys/net/application_layer/uhcp/Makefile` define CFLAGS. (There are also packages that do that, but lets think about packages later).
    * I was also thinking about adding `$(CFLAGS_$(MODULE)/$*.o)` This would allow, for example, adding code coverage `CFLAGS` for some files. 

Some care should also be taken for `CPU` and `BOARD` as they do not define their own module but define `MODULE = cpu` and `MODULE = board`, so only the "right" cpu and module should be included.

And at that point, it should be easy to add dependencies on all Makefiles for all .o files, and maybe even dependency on configuration you give on the command line and then prevent the need for any `make clean`. At the price of rebuilding everything when touching one Makefile but it could be refined in a second pass. (For packages, I would consider okay to need to clean them anyway as we may not have as much control on their build system).

The following step I see, would be changing the `Makefile.dep` to remove this string parsing stuff, and replace it with static dependency definition, but it should take into account optional dependency, features dependency, optional feature providing on optional dependency. And also keep `shell_commands` magic working.


I will try something regarding the 3 points I mentioned, but I would gladly have your feedback on this.
And if you already tried something, know things I missed, know way to tests this correctly, it would help.
